### ISSUE TITLE
Include not resolved with compile_flags.txt if there is a space in the line

### ISSUE THREAD

**Issue by HighCommander4** [clangd/clangd#637]
I don't seem to be able to get the `compile_flags.txt` configuration mechanism to work with a very simple example project.

Given the following workspace structure:

```
test_workspace
|  foo
|  |  header1.hpp
|  |  header2.hpp
|  compile_flags.txt
```

with header1.hpp blank, header2.hpp containing `#include <foo/header1.hpp>`, and `compile_flags.txt` containing `-I .`, the include in header2.hpp is unresolved.

The clangd log shows:

```
I[02:36:45.512] ASTWorker building file /path/to/test_workspace/foo/header1.hpp version 1 with command 
[/path/to/test_workspace]
/path/to/clang/bin/clang-tool -I . /path/to/test_workspace/foo/header1.hpp -fsyntax-only -resource-dir=/path/to/clang/lib/clang/12.0.0
```

The presence of the `-I .` in the command suggests that clangd **is** finding the `compile_flags.txt`, and if I execute that command in the indicated working directory, with `clang-tool` replaced with `clang`, it runs successfully with no diagnostics.

So, I'm not sure why clangd is producing an unresolved inclusion diagnostic.

**Comment by HighCommander4** [comment]
The problem appears to be related to the space in `-I .` -- if I remove the space to get `-I.`, it works.

**Comment by HighCommander4** [comment]
I haven't debugged this fully, but with the space present, [in `clang::ApplyHeaderSearchOptions`](https://searchfox.org/llvm/rev/7df3544e80fb40c742707613cd39ca77f7fea558/clang/lib/Frontend/InitHeaderSearch.cpp#621), the incoming include path entry is stored as `" ."` (with the leading space part of the string), which then cannot be resolved via `vfs::Fileystem::status()`.

**Comment by sam-mccall** [comment]
TL;DR: you need to either omit the space or replace it with a newline - neither compile_flags.txt nor clang itself treat spaces specially, which is why it's treated as part of the filename.

This makes sense as long as you never use a shell, and always run your compilers with exec() :-)

Clang treats argv of `["clang","foo.c","-I","."]` and `["clang","foo.c","-I."]` as equivalent, but doesn't recognize `["clang","foo.c","-I ."]` (or rather as you note it treats it as specifying the include directory " .").

Shell escaping rules say that `clang foo.c -I .` produces the first argv, and `clang foo.c -I.` produces the second argv. `clang foo.c "-I ."` would produce the third, if you wanted that.

compile_flags.txt follows a totally different grammar with no escaping rules: argv is just the file split on newlines. (Okay, it discards empty lines and trims leading/trailing ws, because people complained). In particular, it doesn't know about the structure of arguments (flags, args associated with flags, args joined to flags etc).

This means either of these are valid (argv 1 and 2):
```
-I.
```

```
-I
.
```

while this one produces the bad third argv:

```
-I .
```

**Comment by sam-mccall** [comment]
This is definitely a wart that surprises people, so it's tempting to try to fix it somehow.
But I suspect fixing it would require introducing complicated escaping rules that are just as hard to learn as this wart. (Or making it impossible to express args with spaces in them, which doesn't seem better really).
So maybe we should just document it on the compilation database page?

**Comment by HighCommander4** [comment]
Documenting this is in the CDB docs (and perhaps also in the `compile_flags.txt` section of clangd's "Getting Started" page) would help.

Another thing clangd could do to help is to avoid the (arguably misleading) serialization of `CompileCommand::CommandLine` into a string using [`llvm::join(Inputs.CompileCommand.CommandLine, " ")`](https://searchfox.org/llvm/rev/aebb4a60522615ca4fd542e8ece8df1e4f42d657/clang-tools-extra/clangd/TUScheduler.cpp#647). The invariant that "if I run the compile command that clangd prints and it works, then clangd should work as well" is a useful one. So, if clang is going to interpret the `-I .` as if it were quoted in this case, perhaps it should be serializing it as a quoted string there?

**Comment by sam-mccall** [comment]
Definitely adding shell escaping to the log statement would also be useful. Maybe LLVM already has a function for that that I missed.

**Comment by sam-mccall** [comment]
Going to call this fixed now

**Comment by nico** [comment]
The fix broke tests on windows: http://45.33.8.238/win/32346/step_9.txt

Please take a look, and revert for now if it takes a while to fix.

**Comment by HighCommander4** [comment]
Just wanted to mention for the record this this arguable footgun was propagated to the behaviour of [CompileFlags: Add:](https://clangd.llvm.org/config.html#add) as well (which is used more often than `compile_flags.txt`), such that users who try `Add: [-I path]` (instead of `Add: [-Ipath]` or `Add: [-I, path]`) get tripped up by it as well.

Note sure if that's something we should try and change, or maybe just document in the FAQ.
